### DISCUSSION TITLE
best way to process a large array given a smaller binary mask

### DISCUSSION THREAD

**Discussion by mdrio** [dask/dask#9599] [Q&A] [answered]
Hi, I am working with whole slide images using napari-lazy-openslide.
So I have a large dask array corrisponding to an image with shape (X, Y, C) and a binary mask with a  shape (x, y)  where x <X  and y < Y.

I need to slice the image using the binary mask, process the sliced array, generate a new array with shape (X,Y)  that is zeros everywhere except for the regions corrisponding to the binary mask. 
Ideally, in terms of some pseudo code:

```
out = some_function(image[fit(binary_mask)])
```
where
 ```
image-> shape X,Y, C
binary_mask -> shape x,y
fit(binary_mask) -> shape X,Y,C (it scales array and add a new axis)
out -> shape X,Y
```

I am trying to achieve that this way:

```
def fit(array, h, w):
    x, y = array.shape
    m, n = x * h, y * w
    rescaled = da.broadcast_to(array.reshape(x, 1, y, 1), (x, h, y, w)).reshape(
        m, n
    )
    return da.stack([rescaled, rescaled, rescaled, rescaled], axis=2)

# compute the right h, w...

fitted_mask = fit(binary_mask, h, w)
masked_image = image[fitted_mask]
processed = da.map_blocks(some_func, masked_image, drop_axis=[1])

out = da.zeros(image.shape[:2])
out = da.piecewise(out, [fitted_mask[:, :, 0].rechunk(out.chunksize)], [processed.rechunk(out.chunksize[:1])])

```

Anyway it does not always work, depending on the shapes and the chunks. Probably I misunderstood  how to use da.piecewise. The error that occurs is something like:
```
ValueError: NumPy boolean array indexing assignment cannot assign 26460160 input values to the 0 output values where the mask is true
```

 Moreover, I am not sure this is the right approach from a performance point of view, since the fit function and the rechunks can be heavy tasks.

Any help is appreciated, thanks.

**Comment by ncclementi** [comment, marked as answer]
hi @mdrio,   Dask now has a [Discourse forum](https://dask.discourse.group/t/introduce-yourself/22/23) for usage-questions. We'll see you there! :)

**Reply by mdrio** [reply]
ok!
